src/session.rs

```diff
@@ -370,15 +371,15 @@ impl Session {
                 if !state.is_finished() {
                     bail!(StateProblem::HandshakeNotFinished)
                 } else {
-                    Ok(Session::StatelessTransport(state.try_into()?))
+                    Ok(Session::StatelessTransport((*state).try_into()?))
                 }
             },
             _ => Ok(self)
         }
     }
 }
 
-impl Into<Session> for HandshakeState {
+impl Into<Session> for Box<HandshakeState> {
     fn into(self) -> Session {
         Session::Handshake(self)
```

Comment:
I think it might be better compartmentalization to do

```rust
impl Into<Session> for HandshakeState {
    fn into(self) -> Session {
        Session::Handshake(Box::new(self))
    }
}
```

and then keep the `builder.rs` file unmodified.

Diff looks good otherwise!

Reply:
I wanted to avoid having rustc move (memcpy) the HandshakeState around, which may be more certain when doing the boxing immediately where the struct is initialized, i.e. calling Box::new(HandshakeState { ... }) instead of having the Boxing occur separately in the Into implementation. Do you still think I should change this despite that?

Reply:
I didn't see any noticeable performance difference either way, and a quick check with the lovely `cargo-asm` crate makes it seem like Rust/LLVM is good enough to optimize this problem away!

Reply:
fixed in 64dbb97